controllers/predicates.go

```diff
@@ -73,6 +74,46 @@ func (CertificateRequestPredicate) Update(e event.UpdateEvent) bool {
 	return !reflect.DeepEqual(e.ObjectNew.GetAnnotations(), e.ObjectOld.GetAnnotations())
 }
 
+// Predicate for CertificateSigningRequest changes that should trigger the CertificateSigningRequest reconciler
+//
+// In these cases we want to trigger:
+// - an annotation changed/ was added or removed
+// - a status condition was added or removed
+// - a status condition was changed
+type CertificateSigningRequestPredicate struct {
+	predicate.Funcs
+}
+
+func (CertificateSigningRequestPredicate) Update(e event.UpdateEvent) bool {
+	if e.ObjectOld == nil || e.ObjectNew == nil {
+		// a reference object is missing, just reconcile to be safe
+		return true
+	}
+
+	oldCr, oldOk := e.ObjectOld.(*certificatesv1.CertificateSigningRequest)
+	newCr, newOk := e.ObjectNew.(*certificatesv1.CertificateSigningRequest)
+	if !oldOk || !newOk {
+		// a reference object is invalid, just reconcile to be safe
+		return true
+	}
+
+	if len(oldCr.Status.Conditions) != len(newCr.Status.Conditions) {
+		// fast-fail in case we are certain a non-ready condition was added/ removed
```

Comment:
**Nit**: random tip: if you want to enumerate disjunctive "options" words like "added or removed", use the typography "added/removed". The space after the slash makes the reader think "removed" isn't part of the options.

Ref: https://english.stackexchange.com/questions/548044/how-to-use-a-slash-to-describe-two-options-one-of-which-is-made-up-of-two-or-mo

Reply:
In general, I agree w/ what @maelvls wrote.

I'd personally write `Fail fast` instead of `Fast-fail`:
```suggestion
		// Fail fast in case we are certain a non-ready condition was
		// added or removed.
```